Improve tokenization of identifiers and `k: v` keys

oneeman opened the pull request:
Closes #3113:
- Fix tokenization of `a@b: ` and `A!: `
- Add error for `a@b` (previously tokenized like `a @b`)
- Unify tokenization of upper- and lower-case `k: v` keys
- Fix end-column value for `a: `

Some of the names are admittedly a bit awkward (e.g. for `HasEnding` I couldn't think of a simple term or phrase for "question mark or exclamation point", and some others).  The end-column change wasn't mentioned in the issue, but it was a bug that surfaced when unifying the tokenization of kv keys starting with upper- and lower-case.  Previously, the end-column for `a:` would have been 2 and for `A:` it would have been 3, and I figured they both should be 3.  I changed the value in the test accordingly (from 3 to 4 for `do:`).

josevalim commented on lib/elixir/src/elixir_tokenizer.erl:
```diff
@@ -444,36 +444,30 @@ tokenize([H|_] = String, Line, Column, Scope, Tokens) when ?is_digit(H) ->
       case Rest of
         [$:|T] when ?is_space(hd(T)) ->
           tokenize(T, Line, Column + Length + 1, Scope, [{kw_identifier, {Line, Column, Column + Length + 1}, Atom}|Tokens]);
+        [$:|T] when ?is_atom_start(hd(T)); ?is_digit(hd(T)) ->
```
I think this could be `when hd(T) /= ?:`. We just need to check this clause in case we have `foo::bar`. Otherwise `foo:+` works and it is very weird. :)

josevalim commented on lib/elixir/src/elixir_tokenizer.erl:
```diff
@@ -1003,3 +1015,6 @@ keyword(_) -> false.
   length(atom_to_list(Atom)).
+
+invalid_character_error(Char) ->
+  "invalid character '" ++ [Char] ++ "' in name: ".
```
I think we should rather go with "unexpected token: ", which we already show in other parts of the grammar. :)

josevalim commented:
At first I was not happy about this pull request because you have unified all alias, identifier and atom tokenizing in a single place but I realized that they are directly related so this is most likely the best solution. Great work! I have dropped some comments. Many thanks for PR!

oneeman commented:
Sounds good!  I incorporated the feedback, added tests for the `foo:bar` error since it it had none, and removed some trailing whitespace (still have a long way to go with emacs!).

Will squash and update the commit message if it's ok.

(test failure unrelated to this PR)

josevalim commented:
This looks great!

josevalim commented:
:shipit:

oneeman commented:
Ready :-)

josevalim commented on lib/elixir/test/elixir/kernel/errors_test.exs:
```diff
@@ -37,6 +37,19 @@ defmodule Kernel.ErrorsTest do
+    msg = "nofile:1: keyword argument must be followed by space after: foo:"
+
+    assert_compile_fail SyntaxError, msg, "foo:bar"
+    assert_compile_fail SyntaxError, msg, "foo:+1"
```
Can we add a test for `foo:+` too?

josevalim commented on lib/elixir/test/elixir/kernel/errors_test.exs:
```diff
@@ -37,6 +37,19 @@ defmodule Kernel.ErrorsTest do
   end
 
+  test :invalid_identifier do
+    assert_compile_fail SyntaxError, "nofile:1: unexpected token: foo@", 'foo@'
```
Oops, I was not clear, I think the unexpected token should only be `@`. But now that I think about it, it is better to show the whole thing. Maybe we should go with "syntax error before: foo@`?

josevalim replied:
To be clear, if we use unexpected token, I would expect a single token in the error message. :)

oneeman replied:
Oh, I get it, `foo@` can't be an unexpected token because it's not a valid token.

oneeman commented:
:shipit:

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
